src/main.cpp

```diff
@@ -1083,6 +1083,28 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
                              REJECT_INVALID, "bad-txns-txouttotal-toolarge");
     }
 
+    // Check for non-zero valueBalance when there are no Sapling inputs or outputs
+    if (tx.vShieldedSpend.empty() && tx.vShieldedOutput.empty() && tx.valueBalance != 0) {
```

Comment:
Should we check that this is a Sapling transaction here? Are we sure `valueBalance` will be zero for a sprout transaction? (I don't see where `tx.valueBalance` is defined)

Reply:
`valueBalance` defaults to zero unless otherwise set. Transactions can only enter memory in two ways:

- From the local user via manual transaction creation, which would only set `valueBalance` for a non-Sapling transaction if there was a bug in `zcashd`, and the bug would result in transactions never being successfully submitted to the mempool so it would be detected almost instantly.
- From other peers via the parser, which by definition only sets `valueBalance` for Sapling transactions.

So yes, `valueBalance` will always be zero for non-Sapling transactions.

Reply:
This parser code is after this PR? Cause I didn't find it when I fetch the PR.

Reply:
The parser code was merged several weeks ago, in #3173.